Libraries/Opc.Ua.Server/Subscription/Subscription.cs

```diff
@@ -1965,19 +1965,98 @@ public void ValidateConditionRefresh(OperationContext context)
         /// </summary>
         public void ConditionRefresh()
         {
-            ServerSystemContext systemContext = m_server.DefaultSystemContext.Copy(m_session);
             List<IEventMonitoredItem> monitoredItems = new List<IEventMonitoredItem>();
 
+            lock (m_lock)
+            {
+                // build list of items to refresh.
+                foreach (LinkedListNode<IMonitoredItem> monitoredItem in m_monitoredItems.Values)
+                {
+                    MonitoredItem eventMonitoredItem = monitoredItem.Value as MonitoredItem;
+
+                    if (eventMonitoredItem != null && eventMonitoredItem.EventFilter != null)
+                    {
+                        // add to list that gets reported to the NodeManagers.
+                        monitoredItems.Add(eventMonitoredItem);
+                    }
+                }
+
+                // nothing to do if no event subscriptions.
+                if (monitoredItems.Count == 0)
+                {
+                    return;
+                }
+            }
+
+            ConditionRefresh(monitoredItems, 0);
+        }
+
+
+        /// <summary>
+        /// Refreshes the conditions.
+        /// </summary>
+        public uint ConditionRefresh2(uint monitoredItemId)
+        {
+            uint statusCode = StatusCodes.Good;
+
+            List<IEventMonitoredItem> monitoredItems = new List<IEventMonitoredItem>();
+
+            lock (m_lock)
+            {
+                // build list of items to refresh.
+                if ( m_monitoredItems.ContainsKey(monitoredItemId) )
+                {
+                    LinkedListNode<IMonitoredItem> monitoredItem = m_monitoredItems[monitoredItemId];
+
+                    MonitoredItem eventMonitoredItem = monitoredItem.Value as MonitoredItem;
+
+                    if (eventMonitoredItem != null && eventMonitoredItem.EventFilter != null)
+                    {
+                        // add to list that gets reported to the NodeManagers.
+                        monitoredItems.Add(eventMonitoredItem);
+                    }
+                }
+                else
+                {
+                    return StatusCodes.BadMonitoredItemIdInvalid;
```

Comment:
Is better to throw an exception here, and have the method signature as `public void ConditionRefresh2(uint monitoredItemId)`:
```
throw ServiceResultException.Create(
                  StatusCodes.BadMonitoredItemIdInvalid,
                  "Cannot refresh conditions for a monitored item that does not exist.");
```
Right now the UAMethod call does not return an error for invalid _MonitoredItemId_ because the bad status code is lost at:
  `subscription.ConditionRefresh2( monitoredItemId );`

Reply:
More work is required, as the error code is not getting back to the subscriptionmanager

Reply:
Completed.